### PR TITLE
Rename the repository to a8c-ci-toolkit-buildkite-plugin

### PR DESCRIPTION
**Pull request by rynaardb**
## What it does

- Renames this plugin from `bash-cache-buildkite-plugin` to `a8c-ci-toolkit-buildkite-plugin`
- Updates the `README.md` and `plugin.yml` files

## Additional Improvements Made

- Improve existing tests and another test to check that commands in `bin/` are executable
- ~~Added a new pipeline build step to check that all scripts/commands in `/bin` are executable (`+x` permissions) to safeguard against non-executable files that could be added if one forgot to set the permissions accordingly~~

## How the changes was verified and tested

- All the checks for the [Buildkite Plugin Linter](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/333#01867925-b421-4074-92e1-8703baf414ee) build step passed
- All tests are passing
- ~~The new build step added in this PR was tested and verified working with [this](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/332#01867923-69fd-47ef-b704-12e1b6bb4bfc) and [this](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/333#01867925-b418-4496-bd35-dc45148539a2) build~~

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -1,4 +1,7 @@
 steps:
+  - label: ":terminal: Check Script Permissions"
+    command: .buildkite/commands/check-script-permissions.sh
+
```
Comment (AliSoftware) [inline .buildkite/pipeline.yml]: Nice addition! ❤️

File: plugin.yml [hunk 2]
```diff
@@ -1,7 +1,15 @@
...
-requirements: ['awscli']
+requirements:
+  - aws
+  - gh
+  - jq
+  - node
+  - nvm
+  - ruby
+  - xcodebuild
+  - xcrun
```
Comment (AliSoftware) [inline plugin.yml:12]: 👍

File: tests/environment.bats [hunk 3]
```diff
@@ -2,9 +2,63 @@
...
-  which hash_file > /dev/null
+  which add_host_to_ssh_known_hosts > /dev/null
+  return $?
+}
+
+@test "Exposes 'add_ssh_key_to_agent' command" {
+  source "$PWD/hooks/environment"
+  which add_ssh_key_to_agent > /dev/null
+  return $?
+}
```
Comment (AliSoftware) [inline tests/environment.bats]: I am all for more unit tests, but I can also see us forgetting to add entries to those tests when adding new commands in `bin/` in the future (I for one know I have, and that's also probably the reason why you had to add so many missing ones here).

What do you think about re-writing all those tests about "Exposes x command" into a single test which loops on all commands instead, so that it ends up being a dynamic test with no need for update on new commands?

Something like:
```sh
@test "Exposes all commands in bin/" {
  source "$PWD/hooks/environment"
  for cmd in "$PWD/bin"/*; do
    which "$(basename "$cmd")" > /dev/null
  done
}
```
From what I understand [from the `bats` README](https://github.com/bats-core/bats-core/tree/master#bats-core-bash-automated-testing-system), this should be enough (without the need for `return $?`), as the test will fail as soon as any command has a non-zero exit code:

> Test cases consist of standard shell commands. Bats makes use of Bash's errexit (set -e) option when running test cases. If every command in the test case exits with a 0 status code (success), the test passes. In this way, each line is an assertion of truth.

**Comment by AliSoftware** [comment]
> * All the checks for the [Buildkite Plugin Linter](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/333#01867925-b421-4074-92e1-8703baf414ee) build step passed
> * All tests are passing
> * The new build step added in this PR was tested and verified working with [this](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/332#01867923-69fd-47ef-b704-12e1b6bb4bfc) and [this](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/333#01867925-b418-4496-bd35-dc45148539a2) build

I also wanted to say that I looked at a couple of recent builds **from other repos**, and the renaming of this repo to `a8c-ci-toolkit-buildkite-plugin` didn't seem to have affected them, which is great 👍 

This was more thanks to the fact that the plugin was already cached in the various agents already though, rather than thanks to "GitHub auto-redirect on repo renames" like we thought. I'm sure that, if it were _not_ cached on some agents, then the redirect would still solve this for us anyway like we hoped / mentioned in the Project Thread, but at least even if it didn't, we have the cache helping us on that matter too until we rename the references in all client repos 🙂 

```
  | Running global environment hook | 0s
  | Preparing plugins | 0s
  | # Plugin "github.com/automattic/bash-cache-buildkite-plugin#1.7.0" already checked out (824d1b0)
  | # Plugin "github.com/automattic/git-s3-cache-buildkite-plugin#v1.1.0" already checked out (16d0bea)
  | Running plugin bash-cache environment hook
```

File: tests/environment.bats (hunk 3 above)
Comment (rynaardb) [reply tests/environment.bats]: Thanks! This is a great sugggestion! And tbh, I wasn't aware of the tests either. It was only when I started digging deeper into the repo/plugin that I noticed there were some missing tests for most of the commands.

But I agree with you, I think a single test that iterates over all the commands would be:

- cleaner
- one less thing to remember
- suffient as the test will fail if any of the commands return a non-zero exit code

Comment (rynaardb) [reply tests/environment.bats]: Suggestion applied in 85717fb8c9027de353ea642b2a6171c13f657fcd

File: .buildkite/commands/check-script-permissions.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash -eu
+
+for file in bin/*; do
+  echo "Checking for +x permission on $file..."
+  if ! [ -x "$file" ]; then
+    echo "$file is not executable!"
+    exit 1
+  fi
+done
```
Comment (AliSoftware) [inline .buildkite/commands/check-script-permissions.sh]: Now that I think about it, this would be a good fit to put into the `bats` tests, wdyt?

Like
```sh
@test "Ensure all commands have +x permission" {
  for file in bin/*; do
    [ -x "$file" ]
  done
}
```

**Comment by rynaardb** [comment]
> > * All the checks for the [Buildkite Plugin Linter](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/333#01867925-b421-4074-92e1-8703baf414ee) build step passed
> > * All tests are passing
> > * The new build step added in this PR was tested and verified working with [this](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/332#01867923-69fd-47ef-b704-12e1b6bb4bfc) and [this](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/333#01867925-b418-4496-bd35-dc45148539a2) build
> 
> I also wanted to say that I looked at a couple of recent builds **from other repos**, and the renaming of this repo to `a8c-ci-toolkit-buildkite-plugin` didn't seem to have affected them, which is great 👍
> 
> This was more thanks to the fact that the plugin was already cached in the various agents already though, rather than thanks to "GitHub auto-redirect on repo renames" like we though. I'm sure that, if it were _not_ cached on some agents, then the redirect would still solve this for us anyway like we hoped / mentioned in the Project Thread, but at least even if it didn't, we have the cache helping us on that matter too until we rename the references in all client repos 🙂
> 
> ```
>   | Running global environment hook | 0s
>   | Preparing plugins | 0s
>   | # Plugin "github.com/automattic/bash-cache-buildkite-plugin#1.7.0" already checked out (824d1b0)
>   | # Plugin "github.com/automattic/git-s3-cache-buildkite-plugin#v1.1.0" already checked out (16d0bea)
>   | Running plugin bash-cache environment hook
> ```

Yeah, I'm also really pleased to see that nothing, so far, broke because of the renaming. As far as I can tell, the GitHub repository redirect is also working as expected.

https://github.com/Automattic/bash-cache-buildkite-plugin/ redirects to https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin and with my local repository (still pointing to bash-cache-buildkite-plugin) I'm still able to push & pull without any issues.

File: .buildkite/commands/check-script-permissions.sh (hunk 4 above)
Comment (rynaardb) [reply .buildkite/commands/check-script-permissions.sh]: Sure, we might as well just add another test instead of the build step like I initially did.

Comment (rynaardb) [reply .buildkite/commands/check-script-permissions.sh]: I've replaced the build step with the bats test in 37288d74d80cc2bdf9f0913f02c6de3e6c61aacc

**Comment by AliSoftware** [comment]
I wanted to validate that the tests would properly go red if one of the command indeed didn't have the `+x` flag, so [I cut a `test-pr/39` branch](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/compare/rynaardb/rename-repo...test-pr/39) to `chmod -x` on one of the commands to check.

That's where I noticed that, while the tests were properly going red, [the error output shown in Buildkite logs wasn't super helpful](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/338#01867dcf-13f3-493f-ad43-60dcd0f12b6a) in indicating which one failed (as we're now using a loop with `$cmd`), so I [tweaked the test code a bit](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/commit/af2ae5db65293e559258af612365f0c4a530c91f), to [make the errors hopefully a bit more helpful](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/341#01867df1-97df-4ca1-8085-13a461174077).

Feel free to cherry-pick af2ae5db65293e559258af612365f0c4a530c91f into your branch if you like it 🙂

**Comment by AliSoftware** [comment]
PS: Side note that we'll have to remember to also update the required checks in GitHub's branch protection Settings 🙂 

image

**Comment by rynaardb** [comment]
> PS: Side note that we'll have to remember to also update the required checks in GitHub's branch protection Settings 🙂
> 
> image

Done ✅

**Comment by rynaardb** [comment]
> I wanted to validate that the tests would properly go red if one of the command indeed didn't have the `+x` flag, so [I cut a `test-pr/39` branch](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/compare/rynaardb/rename-repo...test-pr/39) to `chmod -x` on one of the commands to check.
> 
> That's where I noticed that, while the tests were properly going red, [the error output shown in Buildkite logs wasn't super helpful](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/338#01867dcf-13f3-493f-ad43-60dcd0f12b6a) in indicating which one failed (as we're now using a loop with `$cmd`), so I [tweaked the test code a bit](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/commit/af2ae5db65293e559258af612365f0c4a530c91f), to [make the errors hopefully a bit more helpful](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/341#01867df1-97df-4ca1-8085-13a461174077).
> 
> Feel free to cherry-pick [af2ae5d](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/commit/af2ae5db65293e559258af612365f0c4a530c91f) into your branch if you like it 🙂

Thanks, happy to do so! See 04a5952b95d62751efe649275c2a21e7ffaa9b17

**Review by AliSoftware — APPROVED** [review]
Rebased my `test-pr-39` branch on top of your latest commit and could confirm that [the test would still go red if we `chmod -x` one entry and report it in a readable way](https://buildkite.com/automattic/ci-toolkit-buildkite-plugin/builds/345#01867ea9-d1a9-48ca-bbf4-049b9f823b06), so looking great now 👍  🎉  :shipit:
